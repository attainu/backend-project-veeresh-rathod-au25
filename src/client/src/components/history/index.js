import React from 'react';
import PropTypes from 'prop-types';
import times from 'lodash/times';
import map from 'lodash/map';
import { makeStyles, Typography, List } from '@material-ui/core';
import { blue, lime, orange, red, yellow, indigo, cyan, amber, grey } from '@material-ui/core/colors';

import HistoryItem from './history-item';
import { Body } from '../common';

const allColors = [];
times(5, () => allColors.push(
  blue, lime, orange, red, yellow, indigo, cyan, amber
));

const useColors = makeStyles(() => map(allColors, (c) => ({
  color: "white",
  backgroundColor: c[500]
})));

const useStyles = makeStyles(() => ({
  container: {
    marginTop: "1em",
    textAlign: "left"
  },
  title: {
    padding: "8px 16px"
  },
  noData: {
    color: grey[500],
    padding: "16px"
  }
}));

const History = ({ users }) => {
  const colors = useColors();
  const classes = useStyles();

  return (
    <Body isLoggedIn={true}>
      <section className={classes.container}>
        <Typography variant="h4" color="primary" className={classes.title}>
          User History
        </Typography>
        {!users.length ? (
          <Typography variant="h5" className={classes.noData}>No history data</Typography>
        ) : <List>
            {users.map((user, i) => (
              <HistoryItem
                key={user.username}
                color={colors[i]}
                {...user} />)
            )}
          </List>}
      </section>
    </Body>
  );
};

History.propTypes = {
  users: PropTypes.array
};

export default History;