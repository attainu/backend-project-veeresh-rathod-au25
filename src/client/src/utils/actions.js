export const LOGIN_PENDING = "app/auth/LOGIN_STARTED";
export const LOGIN_SUCCESS = "app/auth/LOGIN_SUCCESS";
export const LOGIN_FAILURE = "app/auth/LOGIN_FAILURE";
export const LOGIN_RESET = "app/auth/LOGIN_RESET";

export const LOGOUT_PENDING = "app/auth/LOGOUT_PENDING";
export const LOGOUT_SUCCESS = "app/auth/LOGOUT_SUCCESS";
export const LOGOUT_FAILURE = "app/auth/LOGOUT_FAILURE";
export const LOGOUT_RESET = "app/auth/LOGOUT_RESET";

export const SIGN_UP_PENDING = "app/auth/SIGN_UP_PENDING";
export const SIGN_UP_SUCCESS = "app/auth/SIGN_UP_SUCCESS";
export const SIGN_UP_FAILURE = "app/auth/SIGN_UP_FAILURE";
export const SIGN_UP_RESET = "app/auth/SIGN_UP_RESET";

export const FETCH_ACTIVE_USERS_PENDING = "app/auth/FETCH_ACTIVE_USERS_PENDING";
export const FETCH_ACTIVE_USERS_SUCCESS = "app/auth/FETCH_ACTIVE_USERS_SUCCESS";
export const FETCH_ACTIVE_USERS_FAILURE = "app/auth/FETCH_ACTIVE_USERS_FAILURE";

export const FETCH_USER_HISTORY_PENDING = "app/auth/FETCH_USER_HISTORY_PENDING";
export const FETCH_USER_HISTORY_SUCCESS = "app/auth/FETCH_USER_HISTORY_SUCCESS";
// export const FETCH_USER_HISTORY_FAILURE = "app/auth/FETCH_USER_HISTORY_FAILURE";
