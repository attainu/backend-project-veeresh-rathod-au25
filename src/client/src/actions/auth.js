import get from "lodash/get";
import * as ACTIONS from "../utils/actions";
import { store } from "../App";
import { REGISTRATION_ERROR } from "../utils/constants"
import Socket from "../utils/socket";
import myFetch from "../utils/request";

export const login = async (username, password) => {
  try {
    store.dispatch({ type: ACTIONS.LOGIN_PENDING });
    const { data } = await myFetch("/session/login", {
      method: "POST",
      data: { username, password }
    });
    if (data.status === "OK") {
      Socket.create();
      store.dispatch({ type: ACTIONS.LOGIN_SUCCESS, payload: data.data });
    } else {
      throw data;
    }
  } catch (err) {
    store.dispatch({ type: ACTIONS.LOGIN_FAILURE, payload: err });
  }
};

export const signUp = async (username, password, isAdmin = false) => {
  try {
    store.dispatch({ type: ACTIONS.SIGN_UP_PENDING });
    const { data } = await myFetch("/users/register-user", {
      method: "PUT",
      data: { username, password, isAdmin }
    });
    if (data.status === "OK") {
      store.dispatch({ type: ACTIONS.SIGN_UP_SUCCESS, payload: true });
    } else {
      throw data;
    }
  } catch (err) {
    store.dispatch({
      type: ACTIONS.SIGN_UP_FAILURE,
      payload: get(err, "response.data.message", REGISTRATION_ERROR)
    });
  }
};

export const logout = async () => {
  try {
    store.dispatch({ type: ACTIONS.LOGIN_RESET });
    store.dispatch({ type: ACTIONS.LOGOUT_PENDING });
    const { data } = await myFetch("/session/logout", { method: "GET" });
    if (data.status === "OK") {
      Socket.disconnect();
      store.dispatch({ type: ACTIONS.LOGOUT_SUCCESS, payload: true });
      setTimeout(() => {
        store.dispatch({ type: ACTIONS.LOGOUT_RESET });
      }, 3000);
    } else {
      throw data;
    }
  } catch (err) {
    store.dispatch({ type: ACTIONS.LOGOUT_FAILURE });
  }
};
