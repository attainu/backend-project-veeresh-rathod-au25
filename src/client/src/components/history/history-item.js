import React from 'react';
import PropTypes from 'prop-types';
import capitalize from 'lodash/capitalize';
import { ListItem, ListItemAvatar, ListItemText } from '@material-ui/core';
import UserAvatar from '../home/user-avatar';
import { formatLastVisitTS } from '../../utils';

const HistoryItem = ({ color, username, createdOn, lastVisited, isAdmin }) => {
  return (
    <ListItem>
      <ListItemAvatar>
        <UserAvatar
          initial={username[0]}
          className={color}
          admin={isAdmin} />
      </ListItemAvatar>
      <ListItemText
        primary={capitalize(username)}
        secondary={<em>{`Created on: ${formatLastVisitTS(createdOn)}`}</em>} />
      <ListItemText
        style={{ textAlign: "right" }}
        primary={<em>{`Last online: ${formatLastVisitTS(lastVisited)}`}</em>} />
    </ListItem>
  )
};

HistoryItem.propTypes = {
  username: PropTypes.string,
  lastVisited: PropTypes.string,
  color: PropTypes.string,
  isAdmin: PropTypes.bool
};

export default HistoryItem;