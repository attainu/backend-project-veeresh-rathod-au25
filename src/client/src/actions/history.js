import * as ACTIONS from "../utils/actions";
import { store } from "../App";
import myFetch from "../utils/request";

export const getUserHistory = async () => {
  try {
    store.dispatch({ type: ACTIONS.FETCH_USER_HISTORY_PENDING });
    const { data } = await myFetch("/users/history", { method: "GET" });
    if (data.status === "OK") {
      store.dispatch({ type: ACTIONS.FETCH_USER_HISTORY_SUCCESS, payload: data.data });
    } else {
      throw data;
    }
  } catch (err) {
    store.dispatch({
      type: ACTIONS.FETCH_USER_HISTORY_SUCCESS,
      payload: []
    });
  }
};