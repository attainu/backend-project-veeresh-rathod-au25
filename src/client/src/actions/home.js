import * as ACTIONS from "../utils/actions";
import { store } from "../App";
import Socket from "../utils/socket";

export const getActiveUsers = () => {
  const socket = Socket.getSocket();
  if (socket) {
    store.dispatch({ type: ACTIONS.FETCH_ACTIVE_USERS_PENDING });

    socket.on("activeUsers", (activeUsers) => {
      store.dispatch({ type: ACTIONS.FETCH_ACTIVE_USERS_SUCCESS, payload: activeUsers });
    });
  } else {
    store.dispatch({ type: ACTIONS.FETCH_ACTIVE_USERS_FAILURE, payload: {}, error: "Something went wrong!" });
  }
};
