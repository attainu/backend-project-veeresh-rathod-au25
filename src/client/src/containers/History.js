import React from "react";
import PropTypes from "prop-types";

import { History } from "../components";

import * as Hocs from "../hocs";
import { getUserHistory } from "../actions";

export const HistoryContainer = (props) => {
  React.useEffect(() => {
    getUserHistory();
  }, []);

  return <History {...props} />;
};

HistoryContainer.propTypes = {
  users: PropTypes.array
};

const stateMapper = (state) => ({
  users: state.userHistory.payload || []
});

const HistoryWithAuth = Hocs.withAuth(HistoryContainer);
export default Hocs.withStore(stateMapper)(HistoryWithAuth);