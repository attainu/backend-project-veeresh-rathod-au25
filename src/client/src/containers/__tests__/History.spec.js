jest.unmock("../../actions");

import React from "react";
import { HistoryContainer } from "../History";
import History from "../../components/history";
import { shallow } from "enzyme";
import * as Actions from "../../actions/history";
Actions.getUserHistory = jest.fn();

describe("CONTAINER: src/containers/History.js", () => {
  const defaultProps = { users: [] };
  let wrapper;
  beforeEach(() => {
    React.useEffect = jest.fn().mockImplementation(f => f());
    wrapper = shallow(<HistoryContainer {...defaultProps} />);
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render History component with right props", () => {
    expect(wrapper.find(History)).toBeDefined();
  });

  it("should call getUserHistory action on mount", () => {
    expect(Actions.getUserHistory).toHaveBeenCalled();
  });
});