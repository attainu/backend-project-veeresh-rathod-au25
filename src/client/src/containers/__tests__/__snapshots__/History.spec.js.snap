// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CONTAINER: src/containers/History.js should match snapshot 1`] = `ShallowWrapper {}`;
